operator/internal/controller/vectorized/cluster_controller_test.go

```diff
@@ -0,0 +1,386 @@
+					Port: 8081,
+				},
+			},
+			Resources: vectorizedv1alpha1.RedpandaResourceRequirements{
+				ResourceRequirements: corev1.ResourceRequirements{
+					Limits: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("1"),
+						corev1.ResourceMemory: resource.MustParse("2Gi"),
+					},
+					Requests: corev1.ResourceList{
+						corev1.ResourceCPU:    resource.MustParse("1"),
+						corev1.ResourceMemory: resource.MustParse("2Gi"),
+					},
+				},
+			},
+		},
+	}
+}
+
+func minimalClusterNormalizer(ctx context.Context, testScheme *runtime.Scheme, cluster *vectorizedv1alpha1.Cluster, t *testing.T, ctl *kube.Ctl) []client.Object {
+	var allClusterResources []client.Object
+	var lists []client.ObjectList
+	for gvk := range testScheme.AllKnownTypes() {
+		if strings.HasSuffix(gvk.Kind, "List") {
+			obj, err := testScheme.New(gvk)
+			require.NoError(t, err)
+			if l, ok := obj.(client.ObjectList); ok {
+				lists = append(lists, l)
+			}
+		}
+	}
+
+	clusterSelector := labels.ForCluster(cluster).AsClientSelector()
+	for _, l := range lists {
+		err := ctl.List(ctx, l)
+		if errors.Is(err, &meta.NoKindMatchError{}) {
+			continue
+		}
+		require.NoError(t, err)
+
+		objs, err := meta.ExtractList(l)
+		require.NoError(t, err)
+
+		for _, obj := range objs {
+			o := obj.(client.Object)
+
+			if !clusterSelector.Matches(k8slabels.Set(o.GetLabels())) {
+				continue
+			}
+
+			allClusterResources = append(allClusterResources, o)
+			t.Logf("%T: %s/%s", o, o.GetNamespace(), o.GetName())
+		}
+	}
+
+	for _, rs := range allClusterResources {
+		rs.SetResourceVersion("1")
+		rs.SetUID("1234")
+		rs.SetCreationTimestamp(metav1.Unix(0, 0))
+		rs.SetManagedFields(nil)
+		if ownerRefs := rs.GetOwnerReferences(); len(ownerRefs) > 0 {
+			for i := range ownerRefs {
+				ownerRefs[i].UID = "normalized-uid-1234"
+			}
+			rs.SetOwnerReferences(ownerRefs)
+		}
+
+		if annotations := rs.GetAnnotations(); annotations != nil {
+			annotations["redpanda.com/last-applied"] = "applied"
+		}
+
+		switch obj := rs.(type) {
+		case *corev1.Service:
+			for i := range obj.Spec.Ports {
+				if obj.Spec.Ports[i].NodePort != 0 {
+					obj.Spec.Ports[i].NodePort = 9999
+				}
+			}
+			if obj.Spec.ClusterIP != "" && obj.Spec.ClusterIP != "None" {
+				obj.Spec.ClusterIP = "10.99.99.99"
+			}
+			for i := range obj.Spec.ClusterIPs {
+				if obj.Spec.ClusterIPs[i] != "None" {
+					obj.Spec.ClusterIPs[i] = "10.99.99.99"
+				}
+			}
+
+		case *appsv1.StatefulSet:
+			for i := range obj.Spec.Template.Spec.Containers {
+				for j := range obj.Spec.Template.Spec.Containers[i].Ports {
+					if obj.Spec.Template.Spec.Containers[i].Ports[j].HostPort != 0 {
+						obj.Spec.Template.Spec.Containers[i].Ports[j].HostPort = 9999
+					}
+				}
+			}
+
+			for i := range obj.Spec.Template.Spec.InitContainers {
+				for j := range obj.Spec.Template.Spec.InitContainers[i].Env {
+					if obj.Spec.Template.Spec.InitContainers[i].Env[j].Name == "PROXY_HOST_PORT" {
```

Comment:
Well. That was an absolutely horrific discovery that I wish I had never seen. This does exactly what I was hoping to never see in any codebase. It just rips the allocated nodeport right off the service. Lovely.

Mind adding a comment mentioned as such and/or pointing towards the function / file that causes this to be problematic?